File: packages/kotlin/flappy/src/main/kotlin/flappy/llms/ChatGPT.kt
package flappy.llms

import com.theokanning.openai.client.OpenAiApi
import com.theokanning.openai.completion.chat.ChatCompletionRequest
import com.theokanning.openai.completion.chat.ChatMessage
import com.theokanning.openai.completion.chat.ChatMessageRole
import com.theokanning.openai.service.OpenAiService
import com.theokanning.openai.service.OpenAiService.*
import flappy.*
import okhttp3.OkHttpClient
import java.time.Duration


class ChatGPT(
  private val chatGPTConfig: ChatGPTConfig,
) :
  FlappyLLMBase() {
  override val defaultMaxTokens = -1

  //  https://platform.openai.com/docs/models/overview
  enum class OpenaiModel(val value: String) {
    GPT3("gpt-3.5-turbo-1106"),
    GPT4("gpt-4"),
    GPT4_32K("gpt-4-32k"),
    GPT3_16K("gpt-3.5-turbo-16k")
  }

  private val connectDuration = Duration.ZERO

  private val client: OkHttpClient = defaultClient(chatGPTConfig.token, connectDuration)

  private fun buildOpenAiService(config: ChatGPTConfig): OpenAiService {
    val mapper = defaultObjectMapper()
    val retrofit = if (config.host == null) {
      defaultRetrofit(client, mapper)
    } else {
      defaultRetrofit(client, mapper).newBuilder().baseUrl(config.host).build()
    }

    val api: OpenAiApi = retrofit.create(OpenAiApi::class.java)
    return OpenAiService(api)
  }

  private val service: OpenAiService = buildOpenAiService(chatGPTConfig)

  override fun close() {
    client.dispatcher.executorService.shutdown()
  }

  class ChatGPTConfig @JvmOverloads constructor(
    model: OpenaiModel? = null,
    val token: String,
    val host: String? = null
  ) {
    val chatGPTModel = model ?: OpenaiModel.GPT3
  }

  override suspend fun chatComplete(
    messages: List<FlappyChatMessage>,
    source: String,
    config: LLMGenerateConfig?
  ): LLMResponse {
    logger.info("openai input $messages")

    val chatMessages = messages.map {
      ChatMessage(
        (when (it.role) {
          FlappyRole.ASSISTANT -> ChatMessageRole.ASSISTANT.value()
          FlappyRole.SYSTEM -> ChatMessageRole.SYSTEM.value()
          FlappyRole.USER -> ChatMessageRole.USER.value()
        }),
        it.content
      )
    }

    val chatCompletionRequest =
      ChatCompletionRequest.builder().model(chatGPTConfig.chatGPTModel.value).messages(chatMessages).n(1)
        .topP(config?.topP)
        .temperature(config?.temperature).build()

    val result = service.createChatCompletion(chatCompletionRequest)

    logger.info("openai output $result")

    val choice = result.choices.firstOrNull()

    return if (choice === null)
      LLMResponse.Fail("failed") else
      LLMResponse.Success(choice.message.content)
  }
}

File: packages/nodejs/src/llms/chatgpt.ts
import { LLMBase } from './llm-base'
import { type ChatMLMessage, type ChatMLResponse, type GenerateConfig } from './interface'
import type OpenAI from 'openai'
import { log } from '../utils'
import {
  type ChatCompletionCreateParamsNonStreaming,
  type ChatCompletionCreateParamsBase
} from 'openai/resources/chat/completions'

/**
 * Calculate the default max tokens for a given model.
 * @see https://platform.openai.com/docs/models/overview
 * @param model Model name.
 * @returns
 */
const calcDefaultMaxTokens = (model: ChatGPTModel): number => {
  if (model.includes('16k')) return 16385
  if (model.includes('32k')) return 32768
  if (model.includes('gpt-4')) return 8192
  return 4096
}

// https://platform.openai.com/docs/api-reference/chat/create#chat-create-response_format
const allowJsonResponseFormat = (model: ChatGPTModel): boolean => {
  return model === jsonObjectModel
}

type ChatGPTModel = ChatCompletionCreateParamsBase['model']

export interface ChatGPTOptions {
  model?: ChatGPTModel
  maxTokens?: number
}

const jsonObjectModel: ChatGPTModel = 'gpt-3.5-turbo-1106'

export class ChatGPT extends LLMBase {
  client: OpenAI
  override maxTokens: number
  model: ChatGPTModel
  constructor(client: OpenAI, options?: ChatGPTOptions) {
    const model = options?.model ?? jsonObjectModel
    super()
    this.client = client
    this.model = model
    this.maxTokens = options?.maxTokens ?? calcDefaultMaxTokens(model)
  }

  async chatComplete(messages: ChatMLMessage[], config?: GenerateConfig): Promise<ChatMLResponse> {
    const args: ChatCompletionCreateParamsNonStreaming = {
      model: this.model,
      messages,
      max_tokens: config?.maxTokens,
      ...(allowJsonResponseFormat(this.model) ? { response_format: { type: 'json_object' } } : {}),
      ...config
    }
    const resp = await this.client.chat.completions.create(args)
    const choice = resp.choices[0]
    if (!choice) return { success: false, data: undefined }

    const data = choice.message.content!

    log.debug({ data }, 'openai output')

    return { success: true, data }
  }
}
